linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala

```diff
@@ -777,18 +649,62 @@ private final class IRChecker(unit: LinkingUnit, logger: Logger) {
     }
 
     tree match {
+      // Definitions
+
+      case VarDef(ident, _, vtpe, _, rhs) =>
+        checkDeclareLocalVar(ident)
+        typecheckExpect(rhs, env, vtpe)
+
       // Control flow constructs
 
-      case Block(statsAndExpr) =>
-        val stats :+ expr = statsAndExpr
-        val envAfterStats = stats.foldLeft(env) { (prevEnv, stat) =>
-          typecheckStat(stat, prevEnv)
-        }
-        typecheckExpr(expr, envAfterStats)
+      case Skip() =>
+
+      case Block(trees) =>
+        if (trees.isEmpty)
+          reportError("illegal empty block")
+        typecheckBlockTrees(trees, env)
 
       case Labeled(label, tpe, body) =>
         checkDeclareLabel(label)
-        typecheckExpect(body, env.withLabeledReturnType(label.name, tpe), tpe)
+
+        val expectedType =
+          if (tpe == NoType) AnyType
+          else tpe
+
+        typecheck(body, env.withLabeledReturnType(label.name, expectedType))
```

Comment:
```suggestion
        typecheckExpect(body, env.withLabeledReturnType(label.name, expectedType), tpe)
```
This is important, as the type of the body must also be a subtype of the type of the `Labeled` block.